src/Token/TokenProvider.php

```diff
@@ -0,0 +1,358 @@
+            return $result
+                ->then(function (TokenInterface $token) use ($provider, &$isConstant, &$result) {
+                    // Determine if the token is constant.
+                    if (!$token->getExpiration()) {
+                        $isConstant = true;
+                        return $token;
+                    }
+
+                    // Refresh an expired token.
+                    if (!$token->isExpired()) {
+                        return $token;
+                    }
+                    // Refresh the result and forward the promise.
+                    return $result = $provider($token);
+                })
+                ->otherwise(function($reason) use (&$result) {
+                    // Cleanup rejected promise.
+                    $result = null;
+                    return Promise\Create::promiseFor(null);
+                });
+        };
+    }
+
+    /**
+     * Wraps a token provider and saves provided token in an
+     * instance of Aws\CacheInterface. Forwards calls when no token found
+     * in cache and updates cache with the results.
+     *
+     * @param callable $provider Token provider function to wrap
+     * @param CacheInterface $cache Cache to store the token
+     * @param string|null $cacheKey (optional) Cache key to use
+     *
+     * @return callable
+     */
+    public static function cache(
+        callable $provider,
+        CacheInterface $cache,
+        $cacheKey = null
+    ) {
+        $cacheKey = $cacheKey ?: 'aws_cached_token';
+
+        return function () use ($provider, $cache, $cacheKey) {
+            $found = $cache->get($cacheKey);
+            if ($found instanceof TokenInterface && !$found->isExpired()) {
+                return Promise\Create::promiseFor($found);
+            }
+            return $provider()
+                ->then(function (TokenInterface $token) use (
+                    $cache,
+                    $cacheKey
+                ) {
+                    $cache->set(
+                        $cacheKey,
+                        $token,
+                        null === $token->getExpiration() ?
+                            0 : $token->getExpiration() - time()
+                    );
+
+                    return $token;
+                });
+        };
+    }
+
+    /**
+     * Token provider that retrieves a cached SSO token from the disk
+     *
+     * @return callable
+     */
+    public static function sso($ssoProfileName, $filename = null, $config = [])
+    {
+        $filename = $filename ?: (self::getHomeDir() . '/.aws/config');
+
+        return function () use ($ssoProfileName, $filename, $config) {
+            $ssoProfileName = $ssoProfileName;
+            if (!@is_readable($filename)) {
+                return self::reject("Cannot read token from $filename");
+            }
+            $profiles = self::loadProfiles($filename);
+            if (!isset($profiles[$ssoProfileName])) {
```

Comment:
I might be missing something, but it seems we're checking for a key of `profile <profileName>` here where they're standardized to remove the prefix/space in `self::loadProfiles`.